In [1]:
import vcf

- indels are detected in poly A and poly T sites, or in the TA repeats. 

These issue is addressed here : http://seqanswers.com/forums/showthread.php?t=47721
- so, we don't take them into account

In [4]:
vcf_reader = vcf.Reader(open('./samtools_out/lsd_2_vars.vcf', 'r'))
counter = 0
for record in vcf_reader:
    if "INDEL" in record.INFO.keys():
        print(record)
        print(record.INFO["DP"])
        print(record.INFO["DP4"])
        print(record.INFO)
        print("--------------------")
        counter += 1

print(counter)
    

Record(CHROM=MH893760.2, POS=11566, REF=TTTATTAT, ALT=[TTTAT])
23
[1, 1, 3, 8]
{'INDEL': True, 'IDV': 8, 'IMF': 0.347826, 'DP': 23, 'VDB': 0.579086, 'SGB': -0.676189, 'MQSB': 0.785785, 'MQ0F': 0.173913, 'AF1': 1.0, 'AC1': 1.0, 'DP4': [1, 1, 3, 8], 'MQ': 13, 'FQ': -999.0, 'PV4': [1.0, 1.0, 1.0, 1.0]}
--------------------
Record(CHROM=MH893760.2, POS=12057, REF=GTTTTTTTTTTT, ALT=[GTTTTTTTTTTTTT])
37
[1, 0, 0, 11]
{'INDEL': True, 'IDV': 4, 'IMF': 0.108108, 'DP': 37, 'VDB': 0.118797, 'SGB': -0.676189, 'MQSB': 1.0, 'MQ0F': 0.0810811, 'AF1': 1.0, 'AC1': 1.0, 'DP4': [1, 0, 0, 11], 'MQ': 21, 'FQ': -999.0, 'PV4': [0.0833333, 1.0, 1.0, 1.0]}
--------------------
Record(CHROM=MH893760.2, POS=16010, REF=TAAAAAAAAA, ALT=[TAAAAAAAAAA])
15
[0, 0, 1, 5]
{'INDEL': True, 'IDV': 4, 'IMF': 0.266667, 'DP': 15, 'VDB': 0.258434, 'SGB': -0.616816, 'MQSB': 1.0, 'MQ0F': 0.0, 'AF1': 1.0, 'AC1': 1.0, 'DP4': [0, 0, 1, 5], 'MQ': 33, 'FQ': -999.0}
--------------------
Record(CHROM=MH893760.2, POS=18168, REF=CTTTTTTT

- we take only SNPs.
- if number of reads supporting alternative variant is greater than 10, we print it out

In [5]:
vcf_reader = vcf.Reader(open('./samtools_out/lsd_2_vars_snp.vcf', 'r'))
counter = 0
for record in vcf_reader:
    if int(record.INFO["DP4"][2]) + (record.INFO["DP4"][3]) >= 10:
        print(record)
        print(record.INFO["DP"])
        print(record.INFO["DP4"])
        print(record.INFO)
        print("--------------------")
        counter += 1
print(counter)
    

Record(CHROM=MH893760.2, POS=21695, REF=C, ALT=[T])
88
[12, 7, 30, 27]
{'DP': 88, 'VDB': 0.00825332, 'SGB': -0.693147, 'RPB': 0.0293038, 'MQB': 5.50095e-08, 'MQSB': 0.962405, 'BQB': 0.0458218, 'MQ0F': 0.0681818, 'AF1': 0.510155, 'AC1': 1.0, 'DP4': [12, 7, 30, 27], 'MQ': 36, 'FQ': -999.0, 'PV4': [0.594997, 1.0, 1.0, 1.0]}
--------------------
Record(CHROM=MH893760.2, POS=24453, REF=T, ALT=[C])
20
[1, 3, 6, 10]
{'DP': 20, 'VDB': 0.0529202, 'SGB': -0.689466, 'RPB': 0.366241, 'MQB': 0.0103432, 'MQSB': 0.962269, 'BQB': 0.982301, 'MQ0F': 0.05, 'AF1': 1.0, 'AC1': 1.0, 'DP4': [1, 3, 6, 10], 'MQ': 38, 'FQ': -999.0, 'PV4': [1.0, 0.233036, 1.0, 1.0]}
--------------------
Record(CHROM=MH893760.2, POS=63853, REF=G, ALT=[A])
77
[7, 3, 8, 55]
{'DP': 77, 'VDB': 0.00144445, 'SGB': -0.693147, 'RPB': 0.876526, 'MQB': 2.84408e-06, 'MQSB': 0.0866522, 'BQB': 0.537055, 'MQ0F': 0.012987, 'AF1': 1.0, 'AC1': 1.0, 'DP4': [7, 3, 8, 55], 'MQ': 37, 'FQ': -999.0, 'PV4': [0.000337164, 1.0, 1.0, 1.0]}
----------------

#### those three SNPs showed in IGV:

![21695](./samtools_out/snp_21695.png)

![snp_24453.png](./samtools_out/snp_24453.png)


![snp_63853.png](./samtools_out/snp_63853.png)

In [2]:
%%bash
ls ./input_data/

1_trim_qual_32_1P.fastq
1_trim_qual_32_1U.fastq
1_trim_qual_32_2P.fastq
1_trim_qual_32_2U.fastq
2_trim_crop_35_trail_32_2P.fastq
2_trim_crop_40_trail_32_2P.fastq
2_trim_qual_28_2P.fastq
2_trim_trail_32_2P.fastq
2_trim_win_32_trail_32_2P.fastq
2_trim_win_32_trail_32_crop_30_2P.fastq
LSD-2_S1_L001_R1_001.fastq
LSD-2_S1_L001_R2_001.fastq
over_seq.txt
ref_lsd-3_MH893760.2.fasta
ref_lsd-3_MH893760.2.fasta.fai


In [1]:
%%bash
ls ./samtools_out/

lsd_2_mapped_as_unpaired.bam
lsd_2_mapped_as_unpaired_sorted.bam
lsd_2_mapped_as_unpaired_sorted_rmdup.bam
lsd_2_mapped_as_unpaired_sorted_rmdup.bam.bai
lsd_2_vars.bcf
lsd_2_vars.vcf
lsd_2_vars_snp.vcf
ploid.txt
snp_21695.png
snp_24453.png
snp_63853.png


#### NOTE : we have to redo the calling step, as bcftools can apply to reference only files compressed with bgzip

the command `-Oz -o` means ouput in bzip format 
- we need this compressed format to get consensus (otherwise it doesn't work)
- then we must index compressed `vcf` file, we need it to get derive consensus as well

In [8]:
%%bash
bcftools call \
--consensus-caller \
--variants-only \
--skip-variants indels \
--ploidy-file ./samtools_out/ploid.txt \
./samtools_out/lsd_2_vars.bcf \
-Oz -o ./samtools_out/lsd_2_vars_snp.vcf.gz \


#### we must index the `lsd_2_vars.vcf.gz`

In [9]:
%%bash
bcftools tabix -p vcf ./samtools_out/lsd_2_vars_snp.vcf.gz

### creating consensus

Create consensus sequence by applying VCF variants to a reference fasta file.

- reference sequence
- `bcf` file with variants calls
- indexed `vcf` file

but:
- we didn't call indels, so we don't need to normalize and filter indels

see : http://samtools.github.io/bcftools/howtos/consensus-sequence.html

In [12]:
%%bash
cat ./input_data/ref_lsd-3_MH893760.2.fasta | \
bcftools consensus \
./samtools_out/lsd_2_vars_snp.vcf.gz \
> ./samtools_out/consensus_lsd_2_.fasta

Note: the --sample option not given, applying all records
